### ISSUE TITLE
`install_missing_compilers` attempts to install compilers redundantly when reusing from build cache

### ISSUE THREAD

**Issue by pranav-sivaraman** [LLNL/spack#44417]
### Steps to reproduce

Concretizing the following environment
```yaml
spack:
  specs:
  - gcc@=12.3.0
  view: false
  concretizer:
    unify: true
  config:
    install_tree:
      root: /tmp/spack
      padded_length: 128
    install_missing_compilers: true
  mirrors:
    develop-2024-05-19: https://binaries.spack.io/develop-2024-05-19
  compilers:
  - compiler:
      spec: gcc@=7.5.0
      paths:
        cc: /usr/bin/gcc-7
        cxx: /usr/bin/g++-7
        f77: /usr/bin/gfortran-7
        fc: /usr/bin/gfortran-7
      flags: {}
      operating_system: sles15
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []
``` 

Results in the following spec where everything gets reused.
```console
Input spec
--------------------------------
 -   gcc@12.3.0

Concretized
--------------------------------
 -   gcc@12.3.0%gcc@11.4.0~binutils+bootstrap~graphite~nvptx~piclibs~profiled~strip build_system=autotools build_type=RelWithDebInfo languages='c,c++,fortran' arch=linux-ubuntu22.04-x86_64_v3
 -       ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^gawk@5.3.0%gcc@11.4.0~nls build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64_v3
 -       ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[e]      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3
 -       ^gmp@6.2.1%gcc@11.4.0+cxx build_system=autotools libs=shared,static patches=69ad2e2 arch=linux-ubuntu22.04-x86_64_v3
 -           ^autoconf@2.72%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64_v3
 -       ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-x86_64_v3
 -       ^mpc@1.3.1%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
 -       ^mpfr@4.2.1%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
 -           ^autoconf-archive@2023.02.20%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^perl@5.38.2%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64_v3
 -           ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3
 -           ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3
 -           ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^texinfo@7.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -           ^gettext@0.22.5%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -               ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -               ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-x86_64_v3
 -                   ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-x86_64_v3
 -               ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
 -           ^ncurses@6.5%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-x86_64_v3
 -               ^pkgconf@2.2.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
 -       ^zstd@1.5.6%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64_v3


```

### Error message

However after doing a `spack install` I immediately get the following error. 
```console
==> Error: concretization failed for the following reasons:

   1. gcc compiler '%gcc@7.5.0' incompatible with 'os=ubuntu22.04'
```

### Information on your system

* **Spack:** 0.23.0.dev0 (72deb53832477792b1eff8bd9b6289e84ca85ead)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen3
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
`install_missing_compilers` is not aware of the fact that the original compiler is unnecessary. Not 100% sure if this is worth fixing, cause between now and 0.23 the compiler model will be significantly different.

**Comment by pranav-sivaraman** [comment]
If I remove the `install_missing_compilers` Spack starts fetching from the binary cache, but when installing `gcc` I get the following error.

```console
==> Error: Failed to install gcc due to ProcessError: Command exited with status 1:
    '/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc' 'test.c' '-###'
/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.35' not found (required by /opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc)
/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc)
/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc)
/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_p/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/gcc-12.3.0-ciyytdvvmo4g6bxnnxwz6izkb2b3xing/bin/gcc)
``` 

I guess my question is why Spack reusing a package with a newer `glibc@2.35` when the system version is`glibc@2.31`

**Comment by haampie** [comment]
> I guess my question is why Spack reusing a package with a newer glibc@2.35 when the system version isglibc@2.31

yet another bug :bomb: 

fix in #44428

**Comment by haampie** [comment]
`install_missing_compilers` was deprecated
